File: miniprogram/services/exchange.ts
/**
 * 交换服务
 * 实现交换请求、接受、拒绝等接口
 * 
 * Requirements: 7.1, 7.3, 7.6
 */

import { request } from './request';
import type {
  ExchangeRequest,
  ExchangeResult,
  ExchangeListResult,
  ExchangeHistoryResult,
  ExchangeStats,
  PriceInfo,
} from '../types/exchange';

// 声明 console 对象 (微信小程序环境)
declare const console: {
  log: (...args: any[]) => void;
  error: (...args: any[]) => void;
  warn: (...args: any[]) => void;
};

/**
 * 交换服务类
 * 封装所有交换相关的 API 调用
 */
class ExchangeService {
  // ==================== 交换请求操作 ====================

  /**
   * 创建交换请求
   * 
   * Requirements: 7.1, 7.3
   * @param cardId 卡片 ID
   * @param message 可选的留言
   * @returns 创建的交换请求
   */
  async createExchangeRequest(cardId: string, message?: string): Promise<ExchangeRequest> {
    try {
      const data: Record<string, any> = { card_id: cardId };
      if (message) {
        data.message = message;
      }
      const result = await request.post<ExchangeRequest>('/api/exchanges', data);
      return result;
    } catch (error) {
      console.error('Create exchange request failed:', error);
      throw error;
    }
  }

  /**
   * 接受交换请求
   * 
   * Requirements: 7.6
   * @param exchangeId 交换请求 ID
   * @returns 交换结果
   */
  async acceptExchange(exchangeId: string): Promise<ExchangeResult> {
    try {
      const result = await request.post<ExchangeResult>(`/api/exchanges/${exchangeId}/accept`);
      return result;
    } catch (error) {
      console.error('Accept exchange failed:', error);
      throw error;
    }
  }

  /**
   * 拒绝交换请求
   * 
   * Requirements: 7.6
   * @param exchangeId 交换请求 ID
   */
  async rejectExchange(exchangeId: string): Promise<void> {
    try {
      await request.post<void>(`/api/exchanges/${exchangeId}/reject`);
    } catch (error) {
      console.error('Reject exchange failed:', error);
      throw error;
    }
  }

  /**
   * 取消交换请求（由请求者取消）
   * 
   * @param exchangeId 交换请求 ID
   */
  async cancelExchange(exchangeId: string): Promise<void> {
    try {
      await request.post<void>(`/api/exchanges/${exchangeId}/cancel`);
    } catch (error) {
      console.error('Cancel exchange failed:', error);
      throw error;
    }
  }

  // ==================== 交换请求查询 ====================

  /**
   * 获取待处理的交换请求（收到的）
   * 
   * Requirements: 7.6
   * @returns 待处理请求列表
   */
  async getPendingRequests(): Promise<ExchangeListResult> {
    try {
      const result = await request.get<ExchangeListResult>('/api/exchanges/pending');
      return result;
    } catch (error) {
      console.error('Get pending requests failed:', error);
      throw error;
    }
  }

  /**
   * 获取已发送的交换请求
   * 
   * @returns 已发送请求列表
   */
  async getSentRequests(): Promise<ExchangeListResult> {
    try {
      const result = await request.get<ExchangeListResult>('/api/exchanges/sent');
      return result;
    } catch (error) {
      console.error('Get sent requests failed:', error);
      throw error;
    }
  }

  /**
   * 获取交换历史
   * 
   * @param page 页码
   * @param pageSize 每页数量
   * @returns 交换历史分页结果
   */
  async getExchangeHistory(
    page: number = 1,
    pageSize: number = 20
  ): Promise<ExchangeHistoryResult> {
    try {
      const result = await request.get<ExchangeHistoryResult>('/api/exchanges/history', {
        page,
        page_size: pageSize,
      });
      return result;
    } catch (error) {
      console.error('Get exchange history failed:', error);
      throw error;
    }
  }

  /**
   * 获取交换统计
   * 
   * @returns 交换统计数据
   */
  async getExchangeStats(): Promise<ExchangeStats> {
    try {
      const stats = await request.get<ExchangeStats>('/api/exchanges/stats');
      return stats;
    } catch (error) {
      console.error('Get exchange stats failed:', error);
      throw error;
    }
  }

  // ==================== 价格查询 ====================

  /**
   * 获取交换价格
   * 
   * Requirements: 7.1
   * @param cardId 卡片 ID
   * @returns 价格信息
   */
  async getExchangePrice(cardId: string): Promise<PriceInfo> {
    try {
      const priceInfo = await request.get<PriceInfo>(`/api/exchanges/price/${cardId}`);
      return priceInfo;
    } catch (error) {
      console.error('Get exchange price failed:', error);
      throw error;
    }
  }

  // ==================== 交换详情 ====================

  /**
   * 获取交换请求详情
   * 
   * @param exchangeId 交换请求 ID
   * @returns 交换请求详情
   */
  async getExchangeDetail(exchangeId: string): Promise<ExchangeRequest> {
    try {
      const exchange = await request.get<ExchangeRequest>(`/api/exchanges/${exchangeId}`);
      return exchange;
    } catch (error) {
      console.error('Get exchange detail failed:', error);
      throw error;
    }
  }
}

// 导出单例
export const exchangeService = new ExchangeService();

// 导出类以便测试
export { ExchangeService };

File: life-card-api/src/handlers/exchange.rs
//! Exchange handlers for card exchange operations
//!
//! Requirements: 5.1-5.9, 6.1-6.7

use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde::Deserialize;
use uuid::Uuid;

use crate::error::AppError;
use crate::middleware::auth::{AppState, AuthUserExtractor};
use crate::models::common::{ApiResponse, Pagination};
use crate::models::exchange::CreateExchangeRequestData;
use crate::services::exchange_service::ExchangeService;

// Query Parameters
#[derive(Debug, Deserialize)]
pub struct PaginationQuery {
    #[serde(default = "default_page")]
    pub page: i32,
    #[serde(default = "default_page_size")]
    pub page_size: i32,
}

fn default_page() -> i32 { 1 }
fn default_page_size() -> i32 { 20 }

impl From<PaginationQuery> for Pagination {
    fn from(query: PaginationQuery) -> Self {
        Pagination::new(query.page, query.page_size)
    }
}

// Exchange Request Handlers

/// Create an exchange request
/// 
/// POST /api/exchanges
/// 
/// Requirements: 5.1, 5.2
/// 
/// Request body:
/// ```json
/// {
///     "card_id": "uuid"
/// }
/// ```
pub async fn create_exchange_request(
    State(state): State<AppState>,
    auth: AuthUserExtractor,
    Json(payload): Json<CreateExchangeRequestData>,
) -> Result<impl IntoResponse, AppError> {
    let exchange_service = ExchangeService::new(state.pool.clone(), state.config.clone());
    
    let request = exchange_service
        .create_exchange_request(&auth.user_id, &payload.card_id)
        .await?;

    Ok((StatusCode::CREATED, Json(ApiResponse::success(request))))
}

/// Accept an exchange request
/// 
/// POST /api/exchanges/:exchange_id/accept
/// 
/// Requirements: 5.3
pub async fn accept_exchange(
    State(state): State<AppState>,
    auth: AuthUserExtractor,
    Path(exchange_id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    let exchange_service = ExchangeService::new(state.pool.clone(), state.config.clone());
    
    let result = exchange_service
        .accept_exchange(&exchange_id, &auth.user_id)
        .await?;

    Ok(Json(ApiResponse::success(result)))
}

/// Reject an exchange request
/// 
/// POST /api/exchanges/:exchange_id/reject
/// 
/// Requirements: 5.4
pub async fn reject_exchange(
    State(state): State<AppState>,
    auth: AuthUserExtractor,
    Path(exchange_id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    let exchange_service = ExchangeService::new(state.pool.clone(), state.config.clone());
    
    exchange_service
        .reject_exchange(&exchange_id, &auth.user_id)
        .await?;

    Ok(Json(ApiResponse::<()>::ok_with_message("Exchange request rejected")))
}

/// Cancel an exchange request (by requester)
/// 
/// POST /api/exchanges/:exchange_id/cancel
/// 
/// Requirements: 5.4
pub async fn cancel_exchange(
    State(state): State<AppState>,
    auth: AuthUserExtractor,
    Path(exchange_id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    let exchange_service = ExchangeService::new(state.pool.clone(), state.config.clone());
    
    exchange_service
        .cancel_exchange(&exchange_id, &auth.user_id)
        .await?;

    Ok(Json(ApiResponse::<()>::ok_with_message("Exchange request cancelled")))
}

// Exchange Query Handlers

/// Get pending exchange requests (received as card owner)
/// 
/// GET /api/exchanges/pending
/// 
/// Requirements: 5.6, 5.7
pub async fn get_pending_requests(
    State(state): State<AppState>,
    auth: AuthUserExtractor,
) -> Result<impl IntoResponse, AppError> {
    let exchange_service = ExchangeService::new(state.pool.clone(), state.config.clone());
    
    let requests = exchange_service
        .get_pending_requests(&auth.user_id)
        .await?;

    Ok(Json(ApiResponse::success(requests)))
}

/// Get sent exchange requests (as requester)
/// 
/// GET /api/exchanges/sent
/// 
/// Requirements: 5.6, 5.7
pub async fn get_sent_requests(
    State(state): State<AppState>,
    auth: AuthUserExtractor,
) -> Result<impl IntoResponse, AppError> {
    let exchange_service = ExchangeService::new(state.pool.clone(), state.config.clone());
    
    let requests = exchange_service
        .get_sent_requests(&auth.user_id)
        .await?;

    Ok(Json(ApiResponse::success(requests)))
}

/// Get exchange history (completed exchanges)
/// 
/// GET /api/exchanges/history
/// 
/// Requirements: 5.6, 5.7
pub async fn get_exchange_history(
    State(state): State<AppState>,
    auth: AuthUserExtractor,
    Query(query): Query<PaginationQuery>,
) -> Result<impl IntoResponse, AppError> {
    let exchange_service = ExchangeService::new(state.pool.clone(), state.config.clone());
    
    let history = exchange_service
        .get_exchange_history(&auth.user_id, query.into())
        .await?;

    Ok(Json(ApiResponse::success(history)))
}

// Price Calculation Handler

/// Calculate exchange price for a card
/// 
/// GET /api/exchanges/price/:card_id
/// 
/// Requirements: 6.7
pub async fn get_exchange_price(
    State(state): State<AppState>,
    _auth: AuthUserExtractor,
    Path(card_id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    let exchange_service = ExchangeService::new(state.pool.clone(), state.config.clone());
    
    let price_info = exchange_service
        .calculate_exchange_price(&card_id)
        .await?;

    Ok(Json(ApiResponse::success(price_info)))
}
